File: solutions/clojure/isbn-verifier/1/src/isbn_verifier.clj
(ns isbn-verifier)

(defn only-valid-chars? [str]
  (nil? (re-find #"[^-0-9X]" str))
  )

(defn transcribe [isbn]
  (let [last-idx (dec (count isbn))]
  (filter some? (map (fn[c] (cond
                              (= c 45) nil
                              (= c 88) 10 
                              :else  (- c 48)
                              )) (map int isbn))))
  )

(defn correct-X? [digits]
  (empty? (filter false? (map-indexed #(if (and (== %2 10) (< %1 9)) false true) digits)))
  ) 

(defn isbn? [isbn] ;; <- arglist goes here
  ;; your code goes here
  (if (only-valid-chars? isbn)(
      let [digits (transcribe isbn)] 
           (if (and (= (count digits) 10) (correct-X? digits)) 
             (== 0 (mod (reduce + (map-indexed #(* %2 (- 10 %1)) digits)) 11))
             false
           )
         )             
    false)
)

File: solutions/javascript/isbn-verifier/1/isbn-verifier.js
//
// ISBN-Verifier 

export const isValid = (isbn) => {
  let digits = isbn.split('')
        .filter((d,idx) => d-'0'>=0 || d=='X' )
        .map((d,idx) => 
          {if(d=='X' && idx==9){ return 10;} 
            else return (d-'0')})
        .filter(d => !isNaN(d))
    if (digits.length == 10 && digits.length == 
        isbn.split('').filter(d => d!='-').length)
        return digits.reduce((acc, cur, idx) =>
          acc + cur*(10-idx),0) % 11 == 0;
    else 
      return false
};
